iceoryx_hoofs/include/iceoryx_hoofs/internal/posix_wrapper/mutex.hpp

```diff
@@ -17,81 +17,206 @@
+    /// @todo iox-#1036 remove this, introduced to keep current API temporarily
+    explicit Mutex(const bool f_isRecursive) noexcept;
 
-    /// @attention the destruction of the mutex can fail. This can lead to a program termination!
-    ~mutex() noexcept;
+    /// @brief Destroys the mutex. When the mutex is still locked this will fail and the
+    ///        mutex is leaked! If the MutexThreadTerminationBehavior is set to RELEASE_WHEN_LOCKED
+    ///        a locked mutex is unlocked and the handle is cleaned up correctly.
+    ~Mutex() noexcept;
 
     /// @brief all copy and move assignment methods need to be deleted otherwise
     ///         undefined behavior or race conditions will occure if you copy
     ///         or move mutexe when its possible that they are locked or will
     ///         be locked
-    mutex(const mutex&) = delete;
-    mutex(mutex&&) = delete;
-    mutex& operator=(const mutex&) = delete;
-    mutex& operator=(mutex&&) = delete;
-
-    /// @brief Locks the mutex object and returns true if the underlying c
-    ///         function did not returned any error. If the mutex is already
-    ///         locked the method is blocking till the mutex can be locked.
-    bool lock() noexcept;
-    /// @brief Unlocks the mutex object and returns true if the underlying c
-    ///         function did not returned any error.
-    ///        IMPORTANT! Unlocking and unlocked mutex is undefined behavior
-    ///         and the underlying c function will report success in this case!
-    bool unlock() noexcept;
-
-    /// @brief  Tries to lock the mutex object. If it is not possible to lock
-    ///         the mutex object try_lock will return an error. If the c
-    ///         function fails it will return false, otherwise true.
-    // NOLINTNEXTLINE(readability-identifier-naming) C++ STL code guidelines
-    bool try_lock() noexcept;
-
-    /// @brief  Returns the native handle which then can be used in
-    ///         pthread_mutex_** calls. Required when a pthread_mutex_**
-    ///         call is not abstracted with this wrapper.
-    // NOLINTNEXTLINE(readability-identifier-naming) C++ STL code guidelines
-    pthread_mutex_t get_native_handle() const noexcept;
-
-    // NOLINTNEXTLINE(readability-identifier-naming) C++ STL code guidelines
-    pthread_mutex_t m_handle{};
+    Mutex(const Mutex&) = delete;
+    Mutex(Mutex&&) = delete;
+    Mutex& operator=(const Mutex&) = delete;
+    Mutex& operator=(Mutex&&) = delete;
+
+    /// @brief Locks the mutex.
+    /// @return When it fails it returns a MutexError describing the error.
+    ///         If the mutex is already locked it:
+    ///          * will be non blocking when the lock call comes from the same thread and the
+    ///              mutex type is MutexType::RECURSIVE
+    ///          * will be blocking when for all non MutexType::RECURSIVE
+    ///          * will return a MutexError::DEADLOCK_CONDITION when it is a
```

Comment:
This is actually a great idea. We should do this for all the posix wrappers but I always struggled to implement this since I feared that the amount of error enums is then exploding.

Reply:
I did it and I love it ... thanks for the great idea!